core/silkworm/state/state.hpp

```diff
@@ -17,16 +17,14 @@
 #ifndef SILKWORM_STATE_STATE_HPP_
 #define SILKWORM_STATE_STATE_HPP_
 
-#include <optional>
-
-#include <silkworm/common/base.hpp>
 #include <silkworm/types/account.hpp>
-#include <silkworm/types/block.hpp>
 #include <silkworm/types/receipt.hpp>
 
+#include <silkworm/state/block_state.hpp>
+
 namespace silkworm {
 
-class State {
+class State : public BlockState {
```

Comment:
I don't see any data members in this class, only virtual function, so why is it made non-copyable? Why even have a destructor defined as `=default`?

Reply:
`State` represents Ethereum state, which is typically very big, so you most likely don't want to copy it.

Reply:
The class itself doesn't contain anything, right, so it can be copied.

Reply:
Semantically it should not be copied. At least for now I'd err on the side of caution and prohibit state copying until we get a clear use case for such operation.

Reply:
Hum, it is copied sometimes I think. For example:

![image](https://user-images.githubusercontent.com/1496907/142444426-356034cd-e2ec-4321-b56a-0f09d7382422.png)

allows it to be copied in `ExecutionProcessor` which has the member `IntraBlockState state_;`. So when used in `execute_block(const Block& block, State& state, ...` I believe the state passed to `execute_block` is copied. Not sure if that was intentional.